chainer/functions/normalization/batch_normalization.py

```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
+                gx = (gamma / self.std)[expander] * gy
+        elif cuda.cudnn_enabled and self.use_cudnn and self.cudnn_dim_ok and \
+                self.train:
+            # Note: cuDNN batch normalization backward only works in
+            # "training mode." That is, it does not support
+            # computing gradients in fixed-mean-variance mode, because there
+            # is normally no reason to call backward()
+            # while in test/evaluation mode.
+            dtype = x.dtype
+            handle = cudnn.get_handle()
+            x_desc = cudnn.create_tensor_descriptor(_as4darray(x))
+            derivedBnDesc = cudnn.create_uninitialized_tensor_descriptor()
+            libcudnn.deriveBNTensorDescriptor(derivedBnDesc.value,
+                                              x_desc.value, self.mode)
+            one = numpy.array(1, dtype=dtype).ctypes
+            zero = numpy.array(0, dtype=dtype).ctypes
+            gx = cuda.cupy.empty_like(x)
+            ggamma = cuda.cupy.empty_like(gamma)
+            gbeta = cuda.cupy.empty_like(gamma)
+            libcudnn.batchNormalizationBackward(
+                handle, self.mode, one.data, zero.data,
+                one.data, zero.data, x_desc.value, x.data.ptr,
+                x_desc.value, gy.data.ptr, x_desc.value, gx.data.ptr,
+                derivedBnDesc.value, gamma.data.ptr,
+                ggamma.data.ptr, gbeta.data.ptr,
+                self.eps, self.mean_cache.data.ptr, self.var_cache.data.ptr)
         else:
-            inv_m = numpy.float32(1) / m
-            gx = cuda.elementwise(
-                'T gy, T x_hat, T gamma, T std, T ggamma, T gbeta, T inv_m',
-                'T gx',
-                'gx = (gamma / std) * (gy - (x_hat * ggamma + gbeta) * inv_m)',
-                'bn_bwd')(gy, self.x_hat, gamma[expander], self.std[expander],
-                          ggamma[expander], gbeta[expander], inv_m)
+            gbeta = gy.sum(axis=axis)
+            if self.train:
+                ggamma = (gy * self.x_hat).sum(axis=axis)
+                inv_m = numpy.float32(1) / m
+                gx = cuda.elementwise(
+                    'T gy, T x_hat, T gamma, T std, T ggamma, T gbeta, \
+                    T inv_m',
+                    'T gx',
+                    'gx = (gamma / std) * (gy - (x_hat * ggamma + gbeta) * \
+                    inv_m)',
+                    'bn_bwd')(gy, self.x_hat, gamma[expander],
+                              self.std[expander],
+                              ggamma[expander], gbeta[expander], inv_m)
+            else:
+                # Note: Under normal conditions, this will never be executed
+                # because fixed-mean-variance mode is normally
+                # only used at test/evaluation time.
+                # If cuDNN was called on the forward call, self.x_hat will
+                # be None.
+                if self.x_hat is None:
+                    self.std = xp.sqrt(self.fixed_var,
```

Comment:
How about writing function to calculate `std` and `x_hat`.
There are similar codes.
https://github.com/pfnet/chainer/pull/1310/files#diff-148edea535bb806260c487ca31869a91R116
https://github.com/pfnet/chainer/pull/1310/files#diff-148edea535bb806260c487ca31869a91R179
https://github.com/pfnet/chainer/pull/1310/files#diff-148edea535bb806260c487ca31869a91R220